src/osd/PrimaryLogPG.h

```diff
@@ -1890,5 +1890,40 @@ inline ostream& operator<<(ostream& out,
 void intrusive_ptr_add_ref(PrimaryLogPG::RepGather *repop);
 void intrusive_ptr_release(PrimaryLogPG::RepGather *repop);
 
+#include "common/ceph_crypto.h"
+using ceph::crypto::SHA1;
+#include "rgw/rgw_common.h"
+
+class Fingerprinter {
+public:
+  virtual string do_fingerprint(bufferlist & list) = 0;
+  virtual string fp_to_string(unsigned char *ptr) = 0;
+  virtual ~Fingerprinter() {}
+};
+
+class SHA1_Fingerprinter : public Fingerprinter {
+  unsigned char fingerprint[CEPH_CRYPTO_SHA1_DIGESTSIZE + 1];
+public:
+  string do_fingerprint(bufferlist & list) {
+    char * ptr = list.c_str();
```

Comment:
instead of c_str(), this should iterate over the bufferlist buffers.  it may not alrady be a contiguous buffer and c_str() might trigger an expensive rebuild().

Reply:
BTW this could be added as a bufferlist member, next to crc32c().  I'm not sure that's really where we want it long term, but they two can move together.

Reply:
I agree. I will fix it as your comment.

Reply:
Out of curiosity, why did you choose SHA1 and not XXHASH as the fingerprint?

Reply:
@mykaul SHA1 is just one of option. I'm focusing on dedup operations such as flushing, reference counting. But, as s future work, I will add other hash algorithms.